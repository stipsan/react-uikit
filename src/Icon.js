import { Component, PropTypes } from 'react'
import loader from 'uikit/dist/js/uikit-icons.min.js'

let icons
loader({ icon: { add: (UIkitIcons) => { icons = UIkitIcons } } })

const iconNames = Object.keys(icons)

export default class Icon extends Component {

  static propTypes = {
    icon: PropTypes.oneOf(iconNames).isRequired,
    load: PropTypes.func.isRequired,
    ratio: PropTypes.number.isRequired,
    // @TODO custom viewBox validator
    viewBox: PropTypes.string.isRequired,
  }

  static defaultProps = {
    ratio: 1,
    viewBox: '0 0 20 20',
  }

  render() {
    // @TODO parse icon size from viewbox
    const height = 20
    const width = 20

    const { ratio, viewBox, icon, ...props } = this.props
    const svg = icons[icon] && icons[icon].replace(/^<svg[^>]*>|<\/svg>$/g, '')

    return (
      <span className="uk-icon">
        <svg
          {...props}
          // @TODO optimize
          // eslint-disable-next-line react/no-danger
          dangerouslySetInnerHTML={{ __html: svg }}
          height={height * ratio}
          viewBox={viewBox}
          width={width * ratio}
          xmlns="http://www.w3.org/2000/svg"
        />
      </span>
    )
  }
}
