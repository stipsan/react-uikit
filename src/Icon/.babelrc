{
  "plugins": [
    ["babel-plugin-inline-import", {
      "extensions": [".svg"]
    }]
  ]
}
