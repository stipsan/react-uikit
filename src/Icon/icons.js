import loader from 'uikit/dist/js/uikit-icons.min.js'

import closeIcon from 'uikit/src/images/components/close-icon.svg'
import closeLarge from 'uikit/src/images/components/close-large.svg'
import navbarToggleIcon from 'uikit/src/images/components/navbar-toggle-icon.svg'
import overlayIcon from 'uikit/src/images/components/overlay-icon.svg'
import paginationNext from 'uikit/src/images/components/pagination-next.svg'
import paginationPrevious from 'uikit/src/images/components/pagination-previous.svg'
import searchIcon from 'uikit/src/images/components/search-icon.svg'
import searchLarge from 'uikit/src/images/components/search-large.svg'
import searchNavbar from 'uikit/src/images/components/search-navbar.svg'
import slidenavNext from 'uikit/src/images/components/slidenav-next.svg'
import slidenavNextLarge from 'uikit/src/images/components/slidenav-next-large.svg'
import slidenavPrevious from 'uikit/src/images/components/slidenav-previous.svg'
import slidenavPreviousLarge from 'uikit/src/images/components/slidenav-previous-large.svg'
import spinner from 'uikit/src/images/components/spinner.svg'
import totop from 'uikit/src/images/components/totop.svg'

// base storage for icons
const icons = {}

// Load core icons
loader({ icon: { add: (UIkitIcons) => { Object.assign(icons, UIkitIcons) } } })

// Load component icons
Object.assign(icons, {
  spinner,
  totop,
  'close-icon': closeIcon,
  'close-large': closeLarge,
  'navbar-toggle-icon': navbarToggleIcon,
  'overlay-icon': overlayIcon,
  'pagination-next': paginationNext,
  'pagination-previous': paginationPrevious,
  'search-icon': searchIcon,
  'search-large': searchLarge,
  'search-navbar': searchNavbar,
  'slidenav-next': slidenavNext,
  'slidenav-next-large': slidenavNextLarge,
  'slidenav-previous': slidenavPrevious,
  'slidenav-previous-large': slidenavPreviousLarge,
})

export default icons
