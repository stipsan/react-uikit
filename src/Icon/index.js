import { Component, PropTypes } from 'react'
import loader from 'uikit/dist/js/uikit-icons.min.js'
import cx from 'classnames'
import Spinner from './Spinner'

let icons
loader({ icon: { add: (UIkitIcons) => { icons = UIkitIcons } } })

const iconNames = Object.keys(icons)

export default class Icon extends Component {

  static propTypes = {
    icon: PropTypes.oneOf([...iconNames, 'spinner']).isRequired,
    ratio: PropTypes.number.isRequired,
    // @TODO custom viewBox validator
    viewBox: PropTypes.string.isRequired,
    className: PropTypes.string,
  }

  static defaultProps = {
    className: '',
    ratio: 1,
    viewBox: '0 0 20 20',
  }

  render() {
    // @TODO parse icon size from viewbox
    const height = 20
    const width = 20

    const { className, ratio, viewBox, icon, ...props } = this.props

    if (icon === 'spinner') {
      return <Spinner className={className} />
    }

    const svg = icons[icon] && icons[icon].replace(/^<svg[^>]*>|<\/svg>$/g, '')

    return (
      <span className={cx('uk-icon', className)}>
        <svg
          {...props}
          // @TODO optimize
          // eslint-disable-next-line react/no-danger
          dangerouslySetInnerHTML={{ __html: svg }}
          height={height * ratio}
          viewBox={viewBox}
          width={width * ratio}
          xmlns="http://www.w3.org/2000/svg"
        />
      </span>
    )
  }
}
