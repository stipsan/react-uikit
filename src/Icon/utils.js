
export const getDimensionsFromViewBox = (svg) => {
  const match = svg.match(/view[bB]ox=['"]([\s\S]*?)['"]/m)

  if (!match) return false

  const [, , width, height] = match[1].split(' ')

  return { width, height }
}

export const getInnerSvg = svg => svg.replace(/.*<svg[^>]*>|<\/svg>$/g, '')
