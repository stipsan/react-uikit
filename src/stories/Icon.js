import { storiesOf } from '@kadira/storybook'
import { Icon } from 'uikit-react'

// This is to work around: https://github.com/kadirahq/react-storybook-addon-info/issues/26#issuecomment-229029177
Icon.displayName = 'Icon'

storiesOf('Icon', module)
  .addWithInfo('Basic Usage', '', () => (
    <div className="uk-margin-bottom">
      <Icon icon="check" />
      &nbsp;
      <Icon icon="check" ratio={2} />
      &nbsp;
      <span className="uk-text-danger"><Icon icon="heart" /></span>
    </div>
  ), { header: false, inline: true, propTables: [Icon] })
