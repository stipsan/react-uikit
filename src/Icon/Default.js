import { PureComponent, PropTypes } from 'react'
import cx from 'classnames'

import icons from './icons'
import { getDimensionsFromViewBox, getInnerSvg } from './utils'

export default class Default extends PureComponent {

  static propTypes = {
    component: PropTypes.string.isRequired,
    icon: PropTypes.oneOf(Object.keys(icons)).isRequired,
    ratio: PropTypes.number.isRequired,
    className: PropTypes.string,
  }

  static defaultProps = {
    className: '',
    component: 'span',
    ratio: 1,
  }

  render() {
    const { className, component, ratio, icon, ...props } = this.props
    const svg = icons[icon]
    const { width, height } = getDimensionsFromViewBox(svg)
    const innerSvg = getInnerSvg(svg)

    if (!svg) return false

    const Wrapper = component

    return (
      <Wrapper className={cx('uk-icon', className)} {...props}>
        <svg
          // eslint-disable-next-line react/no-danger
          dangerouslySetInnerHTML={{ __html: innerSvg }}
          height={height * ratio}
          icon={icon}
          viewBox={`0 0 ${width} ${height}`}
          width={width * ratio}
          xmlns="http://www.w3.org/2000/svg"
        />
      </Wrapper>
    )
  }
}
