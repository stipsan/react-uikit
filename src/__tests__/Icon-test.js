import renderer from 'react-test-renderer'
import Icon from '../Icon'

it('should render correctly', () => {
  expect(renderer.create(
    <Icon icon="check" load={jest.fn()} ratio={2} />
  ).toJSON()).toMatchSnapshot()
})
