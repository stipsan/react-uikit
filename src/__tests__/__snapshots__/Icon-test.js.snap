// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<span
  className="uk-icon"
>
  <svg
    dangerouslySetInnerHTML={
      Object {
        "__html": "<polyline fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"1.1\\" points=\\"4,10 8,15 17,4\\"></polyline>",
      }
    }
    height={40}
    viewBox="0 0 20 20"
    width={40}
    xmlns="http://www.w3.org/2000/svg"
  />
</span>
`;

exports[`should render correctly 2`] = `
<Spinner
  className=""
/>
`;
